src/visitors/semantic_analysis_visitor.cpp

```diff
@@ -89,6 +103,99 @@ void SemanticAnalysisVisitor::visit_function_block(const ast::FunctionBlock& nod
     /// -->
 }
 
+void SemanticAnalysisVisitor::visit_statement_block(const ast::StatementBlock& node) {
+    auto last_symbol_table = current_symbol_table;
+    current_symbol_table = node.get_symbol_table();
+    node.visit_children(*this);
+    current_symbol_table = last_symbol_table;
+}
+
+void SemanticAnalysisVisitor::visit_name(const ast::Name& node) {
+    /// <-- This code is a portion of  check 9
+    // There are only two contexts where a random_var is allowed. As the first arg of a random
+    // function or as an item in the RANDOM declaration.
+    // Only the former needs checking.
+    bool ok = true;
+    if (node.is_name()) {  // E.g. for some reason SUFFIX arrives in visit_name
+        auto name = node.get_node_name();
+        auto symbol = current_symbol_table->lookup_in_scope(name);
+        bool is_random_var =
+            symbol ? (symbol->get_properties() == symtab::syminfo::NmodlType::random_var) : false;
+        if (is_random_var) {
+            auto parent = node.get_parent();
+            if (parent) {
+                if (parent->is_var_name()) {
+                    parent = parent->get_parent();
+                    if (parent && parent->is_function_call()) {
+                        // The function must be a random function
+                        auto fname = parent->get_node_name();
+                        if (symtab::syminfo::is_random_function(fname)) {
+                            // but is name the first arg?
+                            ast::FunctionCall* rfun = (ast::FunctionCall*) parent;
```

Comment:
Should this cast be much more elaborate?

Reply:
C++ `dynamic_cast` is sufficient here.